dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h

```diff
@@ -0,0 +1,234 @@
+    TensorShape<ndims> ts;
+    auto e = roi.extent();
+    auto ridx = ndims;
+    ts[--ridx] = e[0] * nchannels;
+    for (size_t idx = 1; idx < ndims; idx++) {
+      ts[--ridx] = e[idx];
+    }
+    ret.emplace_back(ts);
+  }
+
+  return TensorListShape<ndims>{ret};
+}
+
+
+/**
+ * 1. If `rois` is empty, that means whole image is analysed: Roi will the the size of an image
+ * 2. If `rois` is not empty, it is assumed, that Roi is provided for every image in batch.
+ *    In this case, final Roi is an intersection of provided Roi and the image.
+ *    (This is a sanity-check for Rois, that can be larger than image)
+ */
+template <size_t spatial_dims>
+std::vector<Roi<spatial_dims>> AdjustRois(const std::vector<Roi<spatial_dims>> rois,
+                                          const TensorListShape<spatial_dims + 1> &shapes) {
+  assert(rois.empty() || rois.size() == static_cast<size_t>(shapes.num_samples()));
+  std::vector<Roi<spatial_dims>> ret(shapes.num_samples());
+
+  auto whole_image = [](auto shape) -> Roi<spatial_dims> {
+      ivec<spatial_dims> size;
+      for (size_t i = 0; i < spatial_dims; i++)
+        size[i] = shape[spatial_dims - 1 - i];
+      return {0, size};
+  };
+
+  if (rois.empty()) {
+    for (int i = 0; i < shapes.num_samples(); i++) {
+      ret[i] = whole_image(shapes[i]);
+    }
+  } else {
+    for (size_t i = 0; i < rois.size(); i++) {
+      ret[i] = intersection(rois[i], whole_image(shapes[i]));
+    }
+  }
+
+  return ret;
+}
+
+
+/**
+ * Note: Since the brightness-contrast calculation is channel-agnostic (it is performed in the same
+ * way for every channel), SampleDescriptor assumes, that sample is channel-agnostic. Therefore it
+ * needs to be flattened
+ * @see RoiToShape
+ */
+template <class OutputType, class InputType, int ndims>
+std::vector<SampleDescriptor<OutputType, InputType, ndims - 1>>
+CreateSampleDescriptors(const OutListGPU<OutputType, ndims> &out,
+                        const InListGPU<InputType, ndims> &in,
+
+                        const std::vector<float> &brightness, const std::vector<float> &contrast) {
+  std::vector<SampleDescriptor<OutputType, InputType, ndims - 1>> ret(in.num_samples());
+
+  for (int i = 0; i < in.num_samples(); i++) {
+    auto &sample = ret[i];
+    sample.in = in[i].data;
+    sample.out = out[i].data;
+    sample.in_pitch = {};
+    sample.out_pitch = {};
+
+    auto fill_pitch_with_flattening = [](const auto &tv, auto &pitch) {
+        for (size_t i = 0; i < pitch.size(); i++) {
+          pitch[i] = tv.shape[i];
+        }
+        pitch[pitch.size() - 1] *= tv.shape[pitch.size()];
+    };
+
+    fill_pitch_with_flattening(in[i], sample.in_pitch);
+    fill_pitch_with_flattening(out[i], sample.out_pitch);
+
+    sample.brightness = brightness[i];
+    sample.contrast = contrast[i];
+  }
+
+  return ret;
+}
+
+
+template <class OutputType, class InputType, int ndims>
```

Comment:
Would it work for anything other than ndims=2? Looking at the formule, doesn't seem to.
Maybe static_assert?